core/src/fetch_stage.rs

```diff
@@ -6,53 +6,99 @@ use std::{
         Arc,
     },
     thread::{self, Builder, JoinHandle},
+    time::{Duration, Instant},
 };
 
-use crossbeam_channel::{RecvError, RecvTimeoutError};
+use crossbeam_channel::{RecvError, RecvTimeoutError, SendError};
+use solana_metrics::{datapoint_error, datapoint_info};
+use solana_perf::packet::PacketBatch;
 use solana_sdk::packet::{Packet, PacketFlags};
 use solana_streamer::streamer::{PacketBatchReceiver, PacketBatchSender};
 
 #[derive(Debug, thiserror::Error)]
-pub enum Error {
-    #[error("send error")]
-    Send,
-    #[error("recv timeout")]
+pub enum FetchStageError {
+    #[error("send error: {0}")]
+    Send(#[from] SendError<PacketBatch>),
+    #[error("recv timeout: {0}")]
     RecvTimeout(#[from] RecvTimeoutError),
-    #[error("recv error")]
+    #[error("recv error: {0}")]
     Recv(#[from] RecvError),
 }
 
-pub type Result<T> = std::result::Result<T, Error>;
+pub type FetchStageResult<T> = Result<T, FetchStageError>;
 
 pub struct FetchStage {
     thread_hdls: Vec<JoinHandle<()>>,
 }
 
 impl FetchStage {
-    #[allow(clippy::too_many_arguments)]
-    pub fn new_with_sender(
-        exit: &Arc<AtomicBool>,
-        forward_sender: &PacketBatchSender,
-        forward_receiver: PacketBatchReceiver,
+    pub fn new(
+        tpu_forwards_receiver: PacketBatchReceiver,
+        tpu_sender: PacketBatchSender,
+        exit: Arc<AtomicBool>,
     ) -> Self {
-        Self::new_fwd_thread(exit, forward_sender, forward_receiver)
+        let fwd_thread_hdl = Builder::new()
+            .name("fetch_stage-forwarder_thread".to_string())
+            .spawn(move || {
+                let metrics_interval = Duration::from_secs(1);
+                let mut start = Instant::now();
+                let mut tpu_forwards_receiver_max_len = 0usize;
+                let mut tpu_sender_max_len = 0usize;
+                while !exit.load(Ordering::Relaxed) {
+                    match Self::handle_forwarded_packets(&tpu_forwards_receiver, &tpu_sender) {
+                        Ok(()) | Err(FetchStageError::RecvTimeout(RecvTimeoutError::Timeout)) => {}
```

Comment:
looks like fn never returns a timeout, wdyt we rm this from the Error enum type since nothing returns that error here?

Reply:
alternatively should we add a timeout equal to metrics_interval so we don't have no-data on the below metrics when no packets are coming thru?